component/service/src/main/java/org/exoplatform/social/rest/impl/site/SiteRest.java

```diff
@@ -0,0 +1,290 @@
+                           boolean temporalCheck,
+                           @Parameter(description = "to exclude group nodes without page child nodes")
+                           @DefaultValue("false")
+                           @QueryParam("excludeGroupNodesWithoutPageChildNodes")
+                           boolean excludeGroupNodesWithoutPageChildNodes,
+                           @Parameter(description = "to sort with display order")
+                           @DefaultValue("false")
+                           @QueryParam("sortByDisplayOrder")
+                           boolean sortByDisplayOrder,
+                           @Parameter(description = "to filter sites by displayed status")
+                           @DefaultValue("false")
+                           @QueryParam("filterByDisplayed")
+                           boolean filterByDisplayed,
+                           @Parameter(description = "to retrieve sites with displayed status")
+                           @DefaultValue("false")
+                           @QueryParam("displayed")
+                           boolean displayed,
+                           @Parameter(description = "to filter sites by view/edit permissions")
+                           @DefaultValue("false")
+                           @QueryParam("filterByPermissions")
+                           boolean filterByPermission,
+                           @Parameter(description = "Offset of results to retrieve")
+                           @QueryParam("offset")
+                           @DefaultValue("0")
+                           int offset,
+                           @Parameter(description = "Limit of results to retrieve")
+                           @QueryParam("limit")
+                           @DefaultValue("0")
+                           int limit,
+                           @Parameter(description = "Used to retrieve the site label and description in the requested language")
+                           @QueryParam("lang")
+                           String lang) {
+    try {
+      SiteFilter siteFilter = new SiteFilter();
+      if (siteType != null) {
+        siteFilter.setSiteType(SiteType.valueOf(siteType.toUpperCase()));
+      }
+      if (excludedSiteType != null) {
+        siteFilter.setExcludedSiteType(SiteType.valueOf(excludedSiteType.toUpperCase()));
+      }
+      if (excludedSiteName != null) {
+        siteFilter.setExcludedSiteName(excludedSiteName);
+      }
+      siteFilter.setExcludeSpaceSites(excludeSpaceSites);
+      siteFilter.setSortByDisplayOrder(sortByDisplayOrder);
+      if (filterByDisplayed) {
+        siteFilter.setFilterByDisplayed(filterByDisplayed);
+        siteFilter.setDisplayed(displayed);
+      }
+      siteFilter.setLimit(limit);
+      siteFilter.setOffset(offset);
+      List<PortalConfig> sites = layoutService.getSites(siteFilter);
+      return Response.ok(EntityBuilder.buildSiteEntities(sites,
```

Comment:
Shouldn't we use `ETag` to avoid computing the whole response on each page display ? (this way, we send a simple 304 to improve response content download bandwidth + JVM processing)

Reply:
Here we can't implement the Etag because we can not detect changes on all displayed informations. For example, navigation nodes and translated site label/description are calculated with the EntityBuilder and not with layoutService.getSites, so when we add a new navigation node or modify the translation of site label or site description, these changes can not be detected and then the old values still be displayed from the cache until cleaning the cache browser.

Reply:
The requirement is to have a cached response on such huge and repetitive REST call. We always have technical means to make it work, you can see it on Activity Stream by example.